numpy/core/arrayprint.py

```diff
@@ -1596,12 +1597,98 @@ def array_str(a, max_line_width=None, precision=None, suppress_small=None):
         a, max_line_width, precision, suppress_small)
 
 
+def _parse_format_spec(fs):
```

Comment:
It might be worthwhile (and simpler) to use a regex pattern here:
``` python
import re

_PATTERN = re.compile(
    r"([ +-])?"
    r"(\.[0-9]+)?"
    r"([fe])?"
)


def _parse_format_spec(fs):
    match = _PATTERN.fullmatch(fs)
    if match is None:
        raise ValueError("Invalid format specifier")
    sign, precision, typ = match.groups()

    options = {}
    if sign is not None:
        options["sign"] = sign
    if precision is not None:
        options["precision"] = int(precision[1:])
    if typ is not None:
        options["suppress_small"] = typ == "f"
    return options
```

Reply:
You are very right. I don't know how I didn't think about this before. Elegant solution.

Reply:
Now I remember why. The thing with doing the parser "the long way" is that you can throw more specific errors. For example, if they missed the number after the dot, if they typed an unknown field type. I don't know if it's worth the effort. My initial try was to mimic how Python [does it with the built-ins](https://github.com/python/cpython/blob/4aeee0b47b3a2b604bbac37040320ffc88c291f2/Python/formatter_unicode.c#L165).

Do you think it's worth the `ValueError: Format specifier missing precision` or `ValueError: Unknown format code` or do you have any approach to that?

Reply:
> Do you think it's worth the `ValueError: Format specifier missing precision` or `ValueError: Unknown format code` or do you have any approach to that?

You could in principle move some of the validation from the regex matching to the `if` blocks:

``` python
import re

_PATTERN = re.compile(
    r"([ +-])?"
    r"(\.([0-9]+)?)?"
    r"(.|\n)?"
)


def _parse_format_spec(fs):
    match = _PATTERN.fullmatch(fs)
    if match is None:
        raise ValueError("Invalid format specifier")

    sign, period, precision, fmt_code = match.groups()

    options = {}
    if sign is not None:
        options["sign"] = sign
    if period is not None:
        if precision is None:
            raise ValueError("Format specifier missing precision")
        options["precision"] = int(precision)
    if fmt_code is not None:
        if fmt_code not in "fg":
            raise ValueError(f"Unknown format code {fmt_code!r}")
        options["suppress_small"] = fmt_code == "f"
    return options
```